src/storage/mod.rs

```diff
@@ -189,6 +189,120 @@ where
+    pub fn get_mut(&mut self) -> Option<&mut T> {
+        self.storage.get_mut(self.entity)
+    }
+
+    /// Inserts a value into the storage and returns the old one.
+    ///
+    /// Return value can be `None` if the entity passed in was dead.
+    pub fn insert(&mut self, component: T) -> Option<T> {
+        match self.storage.insert(self.entity, component) {
+            InsertResult::Updated(old) => Some(old),
+            _ => None,
+        }
+    }
+
+    /// Removes the component from the storage and returns it.
+    ///
+    /// Return value can be `None` if the entity passed in was dead.
+    pub fn remove(&mut self) -> Option<T> {
+        self.storage.remove(self.entity)
+    }
+}
+
+/// An entry to a storage which does not have a component associated to the entity.
+pub struct VacantEntry<'a, 'b: 'a, T: 'b, D: 'b> {
+    entity: Entity,
+    storage: &'a mut Storage<'b, T, D>,
+}
+
+impl<'a, 'b, T, D> VacantEntry<'a, 'b, T, D>
+    where T: Component,
+          D: DerefMut<Target = MaskedStorage<T>>,
+{
+    /// Inserts a value into the storage and returns the old one.
+    ///
+    /// Return value can be `None` if the entity passed in was dead.
+    pub fn insert(&mut self, component: T) -> Option<T> {
+        match self.storage.insert(self.entity, component) {
+            InsertResult::Updated(old) => Some(old),
+            _ => None,
+        }
+    }
+}
+
+/// Entry to a storage for convenient filling of components or removal based on whether
+/// the entity has a component.
+pub enum StorageEntry<'a, 'b: 'a, T: 'b, D: 'b> {
+    /// Entry variant that is returned if the entity does has a component.
+    Occupied(OccupiedEntry<'a, 'b, T, D>),
+    /// Entry variant that is returned if the entity does not have a component.
+    Vacant(VacantEntry<'a, 'b, T, D>),
+}
+
+impl<'a, 'b, T, D> StorageEntry<'a, 'b, T, D>
+    where T: Component,
+          D: DerefMut<Target = MaskedStorage<T>>
+{
+
+    /// Inserts a component if the entity does not contain a component.
+    ///
+    /// Returns `None` if the entity passed in to get the `Entry` was dead.
+    pub fn or_insert(self, component: T) -> Option<&'a mut T> {
+        match self {
+            StorageEntry::Occupied(occupied) => occupied.storage.get_mut(occupied.entity),
+            StorageEntry::Vacant(vacant) => {
+                let insert_result = vacant.storage.insert(vacant.entity, component);
+                if let InsertResult::EntityIsDead(_) = insert_result {
+                    None
+                }
+                else {
+                    vacant.storage.get_mut(vacant.entity)
+                }
+            }
+        }
+    }
+
+    /// Inserts a component using a default function.
+    ///
+    /// Returns `None` if the entity passed in to get the `Entry` was dead.
+    pub fn or_insert_with<F>(self, component: F) -> Option<&'a mut T>
+        where F: FnOnce() -> T,
+    {
+        self.or_insert(component())
```

Comment:
The whole idea of the method is not to call closure unless component is required.
Instead `or_insert` should be implemented via this one.